onnxruntime/core/framework/session_state.h

```diff
@@ -59,6 +60,93 @@ struct MemoryPatternGroup;
 class MemoryInfo;
 #endif
 
+struct PartialGraphExecutionManager {
+  PartialGraphExecutionManager() {
+    graph_runs_counter_ = 0;
+  }
+
+  int64_t GetNewPartialGraphRunId() {
+    std::lock_guard<OrtMutex> lock(lock_);
+    int64_t new_run_id;
+    if (available_run_ids_.size() > 0) {
+      new_run_id = available_run_ids_.back();
+      available_run_ids_.pop_back();
+    } else {
+      new_run_id = graph_runs_counter_++;
+    }
+
+    ORT_ENFORCE(to_be_executed_runs_.find(new_run_id) == to_be_executed_runs_.end());
+
+    to_be_executed_runs_.insert(new_run_id);
+    return new_run_id;
+  }
+
+  bool IsValidRunId(int64_t run_id) {
+    std::lock_guard<OrtMutex> lock(lock_);
+    auto it_in_progress_runs = graph_runs_.find(run_id);
+    auto it_to_be_executed = to_be_executed_runs_.find(run_id);
+
+    return ((it_in_progress_runs != graph_runs_.end()) ^ (it_to_be_executed != to_be_executed_runs_.end()));
```

Comment:
>^ [](start = 55, length = 1)

It's funny you used an xor op here :joy:

Reply:
@SherlockNoMad `PartialGraphExecutionManager` was designed using the concept of state machine, hence that assert ... if you see run_id is moved between the queues such that it _should not_ co-exist in either queues.

Reply:
how about std::map<int64_t, optional<std::pair<std::shared_ptr<onnxruntime::ExecutionFrame>, size_t>>> 


---
In reply to: [602600185](https://github.com/microsoft/onnxruntime/pull/7077#discussion_r602600185) [](ancestors = 602600185)